lib/auth0/auth0.js

```diff
@@ -0,0 +1,207 @@
+    }
+    if (options.getApiKey && typeof options.getApiKey !== 'function') {
+        throw new Error('The getApiKey option, if specified, must be a function that accepts (ctx, req) and returns an apiKey associated with the request.');
+    }
+    if (options.loginSuccess && typeof options.loginSuccess !== 'function') {
+        throw new Error('The loginSuccess option, if specified, must be a function that accepts (ctx, req, res, baseUrl) and generates a response.');
+    }
+    if (options.loginError && typeof options.loginError !== 'function') {
+        throw new Error('The loginError option, if specified, must be a function that accepts (error, ctx, req, res, baseUrl) and generates a response.');
+    }
+
+    options.clientId = options.clientId || function (ctx, req) {
+        return ctx.secrets.AUTH0_CLIENT_ID;
+    };
+    options.clientSecret = options.clientSecret || function (ctx, req) {
+        return ctx.secrets.AUTH0_CLIENT_SECRET;
+    };
+    options.domain = options.domain || function (ctx, req) {
+        return ctx.secrets.AUTH0_DOMAIN;
+    };
+    options.webtaskSecret = options.webtaskSecret || function (ctx, req) {
+        // By default we don't expect developers to specify WEBTASK_SECRET when
+        // creating authenticated webtasks. In this case we will use webtask token
+        // itself as a JWT signing key. The webtask token of a named webtask is secret
+        // and it contains enough entropy (jti, iat, ca) to pass
+        // for a symmetric key. Using webtask token ensures that the JWT signing secret 
+        // remains constant for the lifetime of the webtask; however regenerating 
+        // the webtask will invalidate previously issued JWTs. 
+        return ctx.secrets.WEBTASK_SECRET || req.x_wt.token;
+    };
+    options.getApiKey = options.getApiKey || function (ctx, req) {
+        if (req.headers.authorization && req.headers.authorization.split(' ')[0] === 'Bearer') {
+            return req.headers.authorization.split(' ')[1];
+        } else if (req.query && req.query.apiKey) {
+            return req.query.apiKey;
+        }
+        return null;
+    };
+    options.loginSuccess = options.loginSuccess || function (ctx, req, res, baseUrl) {
+        res.writeHead(302, { Location: baseUrl + '?apiKey=' + ctx.apiKey });
+        return res.end();
+    };
+    options.loginError = options.loginError || function (error, ctx, req, res, baseUrl) {
+        if (req.method === 'GET') {
+            if (error.redirect) {
+                res.writeHead(302, { Location: error.redirect });
+                return res.end(JSON.stringify(error));
+            }
+            res.writeHead(error.code || 401, { 
+                'Content-Type': 'text/html', 
+                'Cache-Control': 'no-cache' 
+            });
+            return res.end(getNotAuthorizedHtml(baseUrl + '/login'));
```

Comment:
You mean allow people to customize the URL segment name? I am not sure that level of customization is necessary, I am really having hard time coming up with a situation people would really need to change this. We can always add it later if people want to spell "login" in Swahili (which, by the way, is also "login". So no, there are really no scenarios).